src/modm/driver/encoder/encoder.hpp

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2021, Thomas Sommer
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_ENCODER_HPP
+#define MODM_ENCODER_HPP
+
+#include <modm/platform.hpp>
+#include <type_traits>
+
+namespace modm
+{
+/**
+ * \brief	Encoder
+ *
+ * \ingroup	modm_ui_encoder
+ * \author	2021 Thomas Sommer
+ */
+
+/// @ingroup	modm_driver_encoder
+struct encoder
+{
+	typedef enum Division
+	{
+		ONE,
+		TWO,
+		FOUR
+	} Division_t;
+};
+
+template<typename PH_A, typename PH_B, encoder::Division_t DIVISION,
+		 typename UnderlyingType = int8_t>
```

Comment:
Since we have C++20 concepts available you could use `std::signed_integral` to constrain the type instead of the static assert:
```suggestion
		 std::signed_integral UnderlyingType = int8_t>
```

Reply:
Very nice. Didn't know that.